middelware.go

```diff
@@ -309,3 +314,180 @@ func ServiceLinkMiddleware() Middleware {
 		c.Next()
 	}
 }
+
+// selectQValue selects and returns the best value from the allowed set
+// given a header with optional quality values, as you would get for an
+// Accept or Accept-Encoding header. The *first* item in allowed is preferred
+// if there is a tie. If nothing matches, returns an empty string.
+func selectQValue(header string, allowed []string) string {
+	formats := strings.Split(header, ",")
+	best := ""
+	bestQ := 0.0
+	for _, format := range formats {
+		parts := strings.Split(format, ";")
+		name := strings.Trim(parts[0], " \t")
+
+		found := false
+		for _, n := range allowed {
+			if n == name {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			// Skip formats we don't support.
+			continue
+		}
+
+		// Default weight to 1 if no value is passed.
+		q := 1.0
+		if len(parts) > 1 {
+			trimmed := strings.Trim(parts[1], " \t")
+			if strings.HasPrefix(trimmed, "q=") {
+				q, _ = strconv.ParseFloat(trimmed[2:], 64)
+			}
+		}
+
+		// Prefer the first one if there is a tie.
+		if q > bestQ || (q == bestQ && name == allowed[0]) {
+			bestQ = q
+			best = name
+		}
+	}
+
+	return best
+}
+
+type contentEncodingWriter struct {
+	gin.ResponseWriter
+	status      int
+	encoding    string
+	buf         *bytes.Buffer
+	writer      io.Writer
+	minSize     int
+	gzPool      *sync.Pool
+	brPool      *sync.Pool
+	wroteHeader bool
+}
+
+func (w *contentEncodingWriter) Write(data []byte) (int, error) {
+	if w.writer != nil {
+		// We are writing compressed data.
+		return w.writer.Write(data)
+	}
+
+	// Buffer the data until we can decide whether to compress it or not.
+	w.buf.Write(data)
+
+	cl, _ := strconv.Atoi(w.Header().Get("Content-Length"))
```

Comment:
If the `Content-Length` header is missing, I presume the interior of the `Atoi()` function will be `nil`. Does `strconv.Atoi(nil)` need to be handled explicitly before we test `ci` below?

Reply:
Since Go lacks nullable types this will return `0` when missing, which works fine in the check below. That's why we ignore any error it might return, we don't care as we will always check the buffer's length too.